File: java/lib/src/main/java/io/openweb3/xwebhook/EventType.java
package io.openweb3.xwebhook;

import io.openweb3.xwebhook.exceptions.ApiException;
import io.openweb3.xwebhook.internal.api.EventTypeApi;
import io.openweb3.xwebhook.models.EventTypeImportOpenApiIn;
import io.openweb3.xwebhook.models.EventTypeImportOpenApiOut;
import io.openweb3.xwebhook.models.EventTypeIn;
import io.openweb3.xwebhook.models.EventTypeOut;
import io.openweb3.xwebhook.models.EventTypePatch;
import io.openweb3.xwebhook.models.EventTypeUpdate;
import io.openweb3.xwebhook.models.ListResponseEventTypeOut;

public final class EventType {
	private final EventTypeApi api;

	public EventType() {
		api = new EventTypeApi();
	}

	public ListResponseEventTypeOut list(final EventTypeListOptions options) throws ApiException {
		try {
			return api.v1EventTypeList(options.getLimit(), options.getIterator(), null, options.getIncludeArchived(), options.getWithContent());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EventTypeOut create(final EventTypeIn eventTypeIn) throws ApiException {
		return this.create(eventTypeIn, new PostOptions());
	}

	public EventTypeOut create(final EventTypeIn eventTypeIn, final PostOptions options) throws ApiException {
		try {
			return api.v1EventTypeCreate(eventTypeIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EventTypeOut get(final String eventTypeName) throws ApiException {
		try {
			return api.v1EventTypeGet(eventTypeName);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EventTypeOut update(final String eventTypeName, final EventTypeUpdate eventTypeUpdate) throws ApiException {
		try {
			return api.v1EventTypeUpdate(eventTypeName, eventTypeUpdate);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EventTypeOut patch(final String eventTypeName, final EventTypePatch eventTypePatch) throws ApiException {
		try {
			return api.v1EventTypePatch(eventTypeName, eventTypePatch);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void delete(final String eventTypeName) throws ApiException {
		try {
			api.v1EventTypeDelete(eventTypeName, null);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EventTypeImportOpenApiOut importOpenApi(final EventTypeImportOpenApiIn eventTypeImportOpenApiIn) throws ApiException {
		try {
			return this.importOpenApi(eventTypeImportOpenApiIn, new PostOptions());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EventTypeImportOpenApiOut importOpenApi(final EventTypeImportOpenApiIn eventTypeImportOpenApiIn, final PostOptions options) throws ApiException {
		try {
			return api.v1EventTypeImportOpenapi(eventTypeImportOpenApiIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}
}

File: kotlin/lib/src/main/kotlin/EventType.kt
package io.openweb3.xwebhook.kotlin

import io.openweb3.xwebhook.kotlin.exceptions.ApiException
import io.openweb3.xwebhook.kotlin.internal.apis.EventTypeApi
import io.openweb3.xwebhook.kotlin.models.EventTypeImportOpenApiIn
import io.openweb3.xwebhook.kotlin.models.EventTypeImportOpenApiOut
import io.openweb3.xwebhook.kotlin.models.EventTypeIn
import io.openweb3.xwebhook.kotlin.models.EventTypeOut
import io.openweb3.xwebhook.kotlin.models.EventTypePatch
import io.openweb3.xwebhook.kotlin.models.EventTypeUpdate
import io.openweb3.xwebhook.kotlin.models.ListResponseEventTypeOut

class EventType internal constructor(token: String, options: XwebhookOptions) {
    val api = EventTypeApi(options.serverUrl)

    init {
        api.accessToken = token
        api.userAgent = options.getUA()
        options.initialRetryDelayMillis?.let { api.initialRetryDelayMillis = it }
        options.numRetries?.let { api.numRetries = it }
    }

    suspend fun list(options: EventTypeListOptions = EventTypeListOptions()): ListResponseEventTypeOut {
        try {
            return api.v1EventTypeList(options.limit, options.iterator, null, options.includeAchived, options.withContent)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun create(
        eventTypeIn: EventTypeIn,
        options: PostOptions = PostOptions(),
    ): EventTypeOut {
        try {
            return api.v1EventTypeCreate(eventTypeIn, options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun get(eventTypeName: String): EventTypeOut {
        try {
            return api.v1EventTypeGet(eventTypeName)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun update(
        eventTypeName: String,
        eventTypeUpdate: EventTypeUpdate,
    ): EventTypeOut {
        try {
            return api.v1EventTypeUpdate(eventTypeName, eventTypeUpdate)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun patch(
        eventTypeName: String,
        eventTypePatch: EventTypePatch,
    ): EventTypeOut {
        try {
            return api.v1EventTypePatch(eventTypeName, eventTypePatch)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun delete(eventTypeName: String) {
        try {
            api.v1EventTypeDelete(eventTypeName, null)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun importOpenApi(
        eventTypeImportOpenApiIn: EventTypeImportOpenApiIn,
        options: PostOptions = PostOptions(),
    ): EventTypeImportOpenApiOut {
        try {
            return api.v1EventTypeImportOpenapi(eventTypeImportOpenApiIn, options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }
}
